contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+
+    uint premiumInPaymentAsset = totalPremiumInNXM * (tokenPriceInPaymentAsset / 10 ** paymentAssetDecimals);
+    require(premiumInPaymentAsset <= params.maxPremiumInAsset, "Cover: Price exceeds maxPremiumInAsset");
+
+    if (params.payWithNXM) {
+      retrieveNXMPayment(totalPremiumInNXM, params.commissionRatio, params.commissionDestination);
+    } else {
+      retrievePayment(
+        premiumInPaymentAsset,
+        params.paymentAsset,
+        params.commissionRatio,
+        params.commissionDestination
+      );
+    }
+
+    // push the newly created cover
+    _coverData.push(CoverData(
+        params.productId,
+        params.payoutAsset,
+        0 // amountPaidOut
+      ));
+
+    uint coverId = _coverData.length - 1;
+    ICoverNFT(coverNFT).safeMint(params.owner, coverId);
+
+    // Enable this when cover amount tracking is necessary
+    if (coverAmountTrackingEnabled) {
+      totalActiveCoverInAsset[params.payoutAsset] += params.amount;
+    }
+
+    emit CoverBought(coverId, params.productId, 0, msg.sender, params.ipfsData);
+    return coverId;
+  }
+
+  function _buyCover(
+    BuyCoverParams memory params,
+    uint coverId,
+    PoolAllocationRequest[] memory allocationRequests
+  ) internal returns (uint totalPremiumInNXM) {
+
+    // convert to NXM amount
+    uint nxmPriceInPayoutAsset = pool().getTokenPrice(params.payoutAsset);
+    uint remainderAmountInNXM = 0;
+    uint totalCoverAmountInNXM = 0;
+
+    uint _coverSegmentsCount = _coverSegments[coverId].length;
+
+    for (uint i = 0; i < allocationRequests.length; i++) {
+
+      uint requestedCoverAmountInNXM
+        = allocationRequests[i].coverAmountInAsset * 1e18 / nxmPriceInPayoutAsset + remainderAmountInNXM;
+
+      (uint coveredAmountInNXM, uint premiumInNXM, uint rewardsInNXM) = allocateCapacity(
+        params,
+        coverId,
+        stakingPool(allocationRequests[i].poolId),
+        requestedCoverAmountInNXM
+      );
+
+      // apply the global rewards ratio and the total Rewards in NXM
+      tokenController().mintStakingPoolNXMRewards(rewardsInNXM, allocationRequests[i].poolId);
+
+      remainderAmountInNXM = requestedCoverAmountInNXM - coveredAmountInNXM;
+      totalCoverAmountInNXM += coveredAmountInNXM;
+      totalPremiumInNXM += premiumInNXM;
+
+      coverSegmentAllocations[coverId][_coverSegmentsCount].push(
+        PoolAllocation(
+          allocationRequests[i].poolId,
+          SafeUintCast.toUint96(coveredAmountInNXM),
+          SafeUintCast.toUint96(premiumInNXM)
+        )
+      );
+    }
+
+    // priceRatio is normalized on a per year basis (eg. 1.5% per year)
+    uint16 priceRatio = SafeUintCast.toUint16(
+      divRound(
+        totalPremiumInNXM * PRICE_DENOMINATOR * ONE_YEAR / params.period,
+        totalCoverAmountInNXM
+      )
+    );
+
+    _coverSegments[coverId].push(CoverSegment(
+        SafeUintCast.toUint96(totalCoverAmountInNXM * nxmPriceInPayoutAsset / 1e18), // amount
```

Comment:
The current way of calculating the cover amount will likely lead to precision loss, resulting in small cover amount differences that may be annoying (for example 99.999 ETH instead of 100 ETH). A proposed solution: https://github.com/NexusMutual/smart-contracts/commit/0e818ecd38397e606015e40c5ab53e98efda42f2